File: data structures/queue/typescript/queue.ts
class Queue {
	private _items: { [key: number]: any };
	private _length: number;
	private _head: number;
	private _tail: number;

	constructor(...items: any) {
		this._items = {};
		this._length = 0;
		this._head = 0;
		this._tail = 0;
		this.enqueue(...items);
	}

	enqueue(...items: any): Queue {
		for (const item of items) {
			this._items[this._tail++] = item;
			this._length++;
		}

		return this;
	}

	dequeue() {
		if (this.isEmpty) return null;

		const removedItem = this._items[this._head];
		delete this._items[this._head];
		this._length--;
		this._head++;
		return removedItem;
	}

	forEach(callback: Function) {
		let index = 0;
		for (const item of this) {
			callback(item, index++);
		}
	}

	includes(item: any): boolean {
		return Object.values(this._items).includes(item);
	}

	every(callback: Function): boolean {
		let index = 0;
		for (const item of this) {
			if (!callback(item, index++)) return false;
		}

		return true;
	}

	some(callback: Function): boolean {
		let index = 0;
		for (const item of this) {
			if (callback(item, index++)) return true;
		}

		return false;
	}

	clear() {
		this._items = {};
		this._length = 0;
		this._head = 0;
		this._tail = 0;
		return this;
	}

	get peek() {
		return this._items[this._head];
	}
	get isEmpty() {
		return this._length == 0;
	}

	get items() {
		return Object.values(this._items);
	}

	get length() {
		return this._length;
	}

	[Symbol.iterator] = () => {
		let index = this._head;
		return {
			next: () => {
				return {
					value: this._items[index],
					done: index++ == this._tail,
				};
			},
		};
	};
}

export default Queue;

File: data structures/queue/python/queue.py

class Queue:
    def __init__(self, *items) -> None:
        self.__items = {}
        self.__len = 0
        self.__head = 0
        self.__tail = 0
        self.enqueue(*items)

    def enqueue(self, *items):
        for item in items:
            self.__items[self.__tail] = item
            self.__tail += 1
            self.__len += 1
        return self

    def dequeue(self):
        if self.is_empty:
            return

        removed_item = self.__items.pop(self.__head)
        self.__head += 1
        self.__len -= 1
        return removed_item

    def has_item(self, item) -> bool:
        return list(self.__items.values()).count(item) != 0

    def map(self, func):
        for key in self.__items:
            self.__items[key] = func(self.__items[key])
        return self

    def all(self, func) -> bool:
        for item in self:
            if not func(item):
                return False

        return True

    def any(self, func) -> bool:
        for item in self:
            if func(item):
                return True
        return False

    def clear(self):
        self.__items.clear()
        self.__len = 0
        self.__head = 0
        self.__tail = 0
        return self

    @property
    def is_empty(self) -> bool:
        return self.__len == 0

    @property
    def len(self) -> int:
        return self.__len

    def __iter__(self):
        for key in self.__items:
            yield self.__items[key]

    def __str__(self) -> str:
        return f'{list(self.__items.values())}'